dbms/src/Flash/Mpp/ExchangeReceiverCommon.h

```diff
@@ -0,0 +1,429 @@
+                // This can happen when there are old version nodes when upgrading.
+                LOG_ERROR(log, "MPPDataPacket.stream_ids empty, it means ExchangeSender is old version of binary "
+                               "(source_index: {}) while fine grained shuffle of ExchangeReceiver is enabled. "
+                               "Cannot handle this.",
+                          source_index);
+                return false;
+            }
+
+            // packet.stream_ids[i] is corresponding to packet.chunks[i],
+            // indicating which stream_id this chunk belongs to.
+            assert(packet.chunks_size() == packet.stream_ids_size());
+
+            for (int i = 0; i < packet.stream_ids_size(); ++i)
+            {
+                UInt64 stream_id = packet.stream_ids(i) % msg_channels->size();
+                chunks[stream_id].push_back(&packet.chunks(i));
+            }
+        }
+
+        // Still need to send error_ptr or resp_ptr even if packet.chunks_size() is zero.
+        for (size_t i = 0; i < msg_channels->size() && success; ++i)
+        {
+            if (resp_ptr == nullptr && error_ptr == nullptr && chunks[i].empty())
+                continue;
+
+            std::shared_ptr<ReceivedMessage> recv_msg = std::make_shared<ReceivedMessage>(
+                source_index,
+                req_info,
+                tracked_packet,
+                error_ptr,
+                resp_ptr,
+                std::move(chunks[i]));
+            success = (*msg_channels)[i]->push(std::move(recv_msg)) == MPMCQueueResult::OK;
+
+            injectFailPointReceiverPushFail<is_sync>(success);
+
+            // Only the first ExchangeReceiverInputStream need to handle resp.
+            resp_ptr = nullptr;
+        }
+        return success;
+    }
+
+    template <bool is_sync>
+    bool writeNonFineGrain(size_t source_index, const TrackedMppDataPacketPtr & tracked_packet, const mpp::Error * error_ptr, const String * resp_ptr)
+    {
+        bool success = true;
+        auto & packet = tracked_packet->packet;
+        std::vector<const String *> chunks(packet.chunks_size());
+
+        for (int i = 0; i < packet.chunks_size(); ++i)
+            chunks[i] = &packet.chunks(i);
+
+        if (!(resp_ptr == nullptr && error_ptr == nullptr && chunks.empty()))
+        {
+            std::shared_ptr<ReceivedMessage> recv_msg = std::make_shared<ReceivedMessage>(
+                source_index,
+                req_info,
+                tracked_packet,
+                error_ptr,
+                resp_ptr,
+                std::move(chunks));
+
+            success = (*msg_channels)[0]->push(std::move(recv_msg)) == MPMCQueueResult::OK;
+            injectFailPointReceiverPushFail<is_sync>(success);
+        }
+        return success;
+    }
+
+    std::atomic<Int64> * data_size_in_queue;
+    std::vector<MsgChannelPtr> * msg_channels;
+    String req_info;
+    const LoggerPtr log;
+};
+
+class ExchangeReceiverBase
```

Comment:
Rename the file name to `ExchangeReceiverBase`?
And move `ReceiverChannelWriter`  to a separate file